File: advent2017/extras/python/days/day16.py
#!/usr/bin/env python3
"""AoC 2017 Day 16 fancy Python implementation (dance VM + cycle detection)."""

from __future__ import annotations

import argparse
import hashlib
import time
from pathlib import Path

EXPECTED_SHA256 = "6bb64ef97ccf665f21eccff0a7045717f0a03d39ae06aaac5495dd6fff650818"
EXPECTED_PART1 = "kgdchlfniambejop"
EXPECTED_PART2 = "fjpmholcibdgeakn"


def default_input_path() -> Path:
    cands = [
        Path("advent2017/Day16/d16_input.txt"),
        Path("Day16/d16_input.txt"),
        Path("../Day16/d16_input.txt"),
        Path("../../Day16/d16_input.txt"),
        Path(__file__).resolve().parents[2] / "Day16" / "d16_input.txt",
    ]
    for c in cands:
        if c.exists():
            return c.resolve()
    raise FileNotFoundError("input not found for day 16")


def sha256_file(path: Path) -> str:
    h = hashlib.sha256()
    with path.open("rb") as f:
        for chunk in iter(lambda: f.read(1 << 20), b""):
            h.update(chunk)
    return h.hexdigest()


def parse_moves(raw: str) -> list[str]:
    return [m.strip() for m in raw.strip().split(",") if m.strip()]


def apply_once(state: list[str], moves: list[str]) -> None:
    for mv in moves:
        t = mv[0]
        arg = mv[1:]
        if t == "s":
            x = int(arg)
            state[:] = state[-x:] + state[:-x]
        elif t == "x":
            a, b = map(int, arg.split("/"))
            state[a], state[b] = state[b], state[a]
        elif t == "p":
            a, b = arg.split("/")
            ia = state.index(a)
            ib = state.index(b)
            state[ia], state[ib] = state[ib], state[ia]
        else:
            raise ValueError("bad move")


def dance(moves: list[str], rounds: int) -> str:
    state = list("abcdefghijklmnop")
    seen: dict[str, int] = {}

    i = 0
    while i < rounds:
        key = "".join(state)
        if key in seen:
            cycle = i - seen[key]
            rem = (rounds - i) % cycle
            for _ in range(rem):
                apply_once(state, moves)
            return "".join(state)
        seen[key] = i
        apply_once(state, moves)
        i += 1

    return "".join(state)


def main() -> int:
    ap = argparse.ArgumentParser(description="AoC 2017 Day 16 fancy Python")
    ap.add_argument("--part", type=int, required=True, choices=[1, 2])
    ap.add_argument("--input")
    args = ap.parse_args()

    input_path = Path(args.input).resolve() if args.input else default_input_path()
    got_sha = sha256_file(input_path)
    if got_sha != EXPECTED_SHA256:
        raise SystemExit(f"checksum mismatch: {got_sha} != {EXPECTED_SHA256}")

    moves = parse_moves(input_path.read_text(encoding="utf-8"))

    t0 = time.perf_counter_ns()
    answer = dance(moves, 1 if args.part == 1 else 1_000_000_000)
    runtime_ms = (time.perf_counter_ns() - t0) / 1e6

    expected = EXPECTED_PART1 if args.part == 1 else EXPECTED_PART2
    if answer != expected:
        raise SystemExit(f"answer mismatch for part {args.part}: got {answer!r}, expected {expected!r}")

    print(answer)
    print(f"[python-fancy] day=16 part={args.part} runtime_ms={runtime_ms:.3f}", file=__import__("sys").stderr)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: advent2017/extras/rust/src/bin/day16.rs
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::env;
use std::fs;
use std::path::Path;
use std::time::Instant;

const EXPECTED_SHA: &str = "6bb64ef97ccf665f21eccff0a7045717f0a03d39ae06aaac5495dd6fff650818";
const EXPECTED_PART1: &str = "kgdchlfniambejop";
const EXPECTED_PART2: &str = "fjpmholcibdgeakn";

fn resolve_input(provided: &str) -> String {
    if !provided.is_empty() { return provided.to_string(); }
    let cands = ["advent2017/Day16/d16_input.txt", "Day16/d16_input.txt", "../Day16/d16_input.txt", "../../Day16/d16_input.txt"];
    for c in cands { if Path::new(c).exists() { return c.to_string(); } }
    panic!("input not found");
}

fn sha256_file(path: &str) -> String {
    let b = fs::read(path).expect("read");
    let mut h = Sha256::new(); h.update(&b); hex::encode(h.finalize())
}

fn parse_moves(raw: &str) -> Vec<String> {
    raw.trim().split(',').map(|x| x.trim().to_string()).filter(|x| !x.is_empty()).collect()
}

fn apply_once(state: &mut Vec<u8>, moves: &[String]) {
    for mv in moves {
        let t = mv.as_bytes()[0] as char;
        let arg = &mv[1..];
        match t {
            's' => {
                let x: usize = arg.parse().expect("int");
                let n = state.len();
                let mut tmp = state[(n - x)..].to_vec();
                tmp.extend_from_slice(&state[..(n - x)]);
                *state = tmp;
            }
            'x' => {
                let p = arg.split('/').collect::<Vec<&str>>();
                let a: usize = p[0].parse().expect("int");
                let b: usize = p[1].parse().expect("int");
                state.swap(a, b);
            }
            'p' => {
                let p = arg.split('/').collect::<Vec<&str>>();
                let a = p[0].as_bytes()[0];
                let b = p[1].as_bytes()[0];
                let ia = state.iter().position(|&c| c == a).unwrap();
                let ib = state.iter().position(|&c| c == b).unwrap();
                state.swap(ia, ib);
            }
            _ => panic!("bad move"),
        }
    }
}

fn dance(moves: &[String], rounds: usize) -> String {
    let mut state = b"abcdefghijklmnop".to_vec();
    let mut seen: HashMap<String, usize> = HashMap::new();
    let mut i = 0usize;

    while i < rounds {
        let key = String::from_utf8(state.clone()).unwrap();
        if let Some(prev) = seen.get(&key) {
            let cycle = i - prev;
            let rem = (rounds - i) % cycle;
            for _ in 0..rem { apply_once(&mut state, moves); }
            return String::from_utf8(state).unwrap();
        }
        seen.insert(key, i);
        apply_once(&mut state, moves);
        i += 1;
    }

    String::from_utf8(state).unwrap()
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut part = 0;
    let mut input = String::new();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--part" => { part = args[i + 1].parse().unwrap(); i += 2; }
            "--input" => { input = args[i + 1].clone(); i += 2; }
            _ => panic!("unknown arg"),
        }
    }
    if part != 1 && part != 2 { panic!("--part 1|2 required"); }

    let in_path = resolve_input(&input);
    if sha256_file(&in_path) != EXPECTED_SHA { panic!("checksum mismatch"); }
    let moves = parse_moves(&fs::read_to_string(&in_path).expect("read_to_string"));

    let t0 = Instant::now();
    let ans = if part == 1 { dance(&moves, 1) } else { dance(&moves, 1_000_000_000) };
    let expected = if part == 1 { EXPECTED_PART1 } else { EXPECTED_PART2 };
    if ans != expected { panic!("answer mismatch"); }

    println!("{}", ans);
    let ms = t0.elapsed().as_nanos() as f64 / 1e6;
    eprintln!("[rust-fancy] day=16 part={} runtime_ms={:.3}", part, ms);
}
